src/DataStructures/Tensor/Expressions/Evaluate.hpp

```diff
@@ -90,18 +90,110 @@ auto evaluate(const T& rhs_te) {
+          Requires<Tensor_detail::is_tensor<T>::value> = nullptr,
+          Requires<std::is_base_of<Expression, TE>::value> = nullptr>
+void evaluate(const gsl::not_null<T*> lhs_tensor, const TE& rhs_te) {
+  using lhs_tensorindex_list =
+      tmpl::list<std::decay_t<decltype(LhsTensorIndices)>...>;
+  using rhs_tensorindex_list = typename TE::args_list;
+  static_assert(
+      tmpl::equal_members<lhs_tensorindex_list, rhs_tensorindex_list>::value,
+      "The generic indices on the LHS of a tensor equation (that is, the "
+      "template parameters specified in evaluate<...>) must match the generic "
+      "indices of the RHS TensorExpression. This error occurs as a result of a "
+      "call like evaluate<ti_a, ti_b>(R(ti_A, ti_b) * S(ti_a, ti_c)), where "
+      "the generic indices of the evaluated RHS expression are ti_b and ti_c, "
+      "but the generic indices provided for the LHS are ti_a and ti_b.");
+  static_assert(
+      tmpl::is_set<std::decay_t<decltype(LhsTensorIndices)>...>::value,
+      "Cannot evaluate a tensor expression to a LHS tensor with a repeated "
+      "generic index, e.g. evaluate<ti_a, ti_a>.");
+  static_assert(
+      not detail::contains_indices_to_contract<sizeof...(LhsTensorIndices)>(
+          {{std::decay_t<decltype(LhsTensorIndices)>::value...}}),
+      "Cannot evaluate a tensor expression to a LHS tensor with generic "
+      "indices that would be contracted, e.g. evaluate<ti_A, ti_a>.");
+  using rhs_symmetry = typename TE::symmetry;
+  using rhs_tensorindextype_list = typename TE::index_list;
+
+  // Stores (potentially reordered) symmetry and indices expected for the LHS
+  // tensor, with index order specified by LhsTensorIndices
+  using lhs_tensor_symm_and_indices =
+      LhsTensorSymmAndIndices<rhs_tensorindex_list, lhs_tensorindex_list,
+                              rhs_symmetry, rhs_tensorindextype_list>;
+  // Instead of simply checking that `T` is the correct LHS Tensor type,
+  // individual checks for the data type, symmetry, and index_list are carried
+  // out to provide more fine-grained feedback
+  static_assert(
+      std::is_same_v<typename T::type, typename TE::type>,
+      "The data type stored by the LHS tensor does not match the data type "
+      "stored by the RHS expression.");
+  static_assert(
+      std::is_same_v<typename T::symmetry,
+                     typename lhs_tensor_symm_and_indices::symmetry>,
+      "The symmetry of the LHS tensor does not match the symmetry of the "
+      "evaluated RHS expression.");
+  static_assert(
+      std::is_same_v<
+          typename T::index_list,
+          typename lhs_tensor_symm_and_indices::tensorindextype_list>,
+      "The index list of the LHS tensor does not match the index list of the "
+      "evaluated RHS expression.");
+
+  // Fill LHS tensor
+  for (size_t i = 0; i < T::size(); i++) {
+    (*lhs_tensor)[i] =
+        rhs_te.template get<typename T::structure,
```

Comment:
Making `Tensor`'s `structure` alias public was necessary to be able to use it outside the `Tensor` class here. If that change to `Tensor` is unwanted/problematic, let me know, and I can instead factor out this loop into a new small helper function in the `Tensor` class.